File: ext/RmlUi.Native/src/Rml.cpp
#include "RmlNative.h"
#include "RmlUi/Core.h"
#include "EventListener.h"
#include <string>

static std::unordered_map<Rml::String*, Rml::String> dataModelPool;

RMLUI_CAPI bool rml_Initialise() {
    return Rml::Initialise();
}

RMLUI_CAPI void rml_Shutdown() {
    Rml::Shutdown();
}

RMLUI_CAPI void rml_SetRenderInterface(Rml::RenderInterface *renderInterface) {
    Rml::SetRenderInterface(renderInterface);
}

RMLUI_CAPI void rml_SetSystemInterface(Rml::SystemInterface *systemInterface) {
    Rml::SetSystemInterface(systemInterface);
}

RMLUI_CAPI void rml_SetFileInterface(Rml::FileInterface* fileInterface) {
    Rml::SetFileInterface(fileInterface);
}

RMLUI_CAPI void rml_LoadFontFace(const char *fileName, bool fallbackFace, Rml::Style::FontWeight weight) {
    Rml::LoadFontFace(fileName, fallbackFace, weight);
}

RMLUI_CAPI Rml::EventId rml_RegisterEventType(const char* type, bool interruptible, bool bubbles, Rml::DefaultActionPhase default_action_phase) {
    return Rml::RegisterEventType(type, interruptible, bubbles, default_action_phase);
}

RMLUI_CAPI Rml::EventListener *rml_CreateEventListener(::onProcessEvent onProcessEvent, ::onAttachEvent onAttachEvent, ::onDetachEvent onDetachEvent) {
    return new EventListenerProxy(onProcessEvent, onAttachEvent, onDetachEvent);
}

RMLUI_CAPI void rml_ReleaseEventListener(Rml::EventListener *eventListener) {
    delete eventListener;
}

RMLUI_CAPI void rml_RemoveContext(const char *name) {
    Rml::RemoveContext(name);
}

RMLUI_CAPI void rml_RegisterPlugin(Rml::Plugin *plugin) {
    Rml::RegisterPlugin(plugin);
}

RMLUI_CAPI Rml::RenderInterface* rml_GetRenderInterface() {
    return Rml::GetRenderInterface();
}

RMLUI_CAPI Rml::SystemInterface* rml_GetSystemInterface() {
    return Rml::GetSystemInterface();
}

RMLUI_CAPI std::string* rml_CreateString(const char* value) {
    auto str = new Rml::String(value);
    dataModelPool[str] = *str;
    return str;
}

RMLUI_CAPI void rml_Log(Rml::Log::Type type,  char* message) {
    Rml::Log::Message(type, message);
}

RMLUI_CAPI void rml_UpdateString(std::string* str, const char* new_value) {
    *str = new_value;
}

RMLUI_CAPI void rml_FreeString(std::string* str) {
        dataModelPool.erase(str);
}

File: ext/RmlUI.Net/Rml.cs
﻿using System;

namespace RmlUiNet
{
    public static class Rml
    {
        public static bool Initialise()
        {
            return Native.Rml.Initialise();
        }

        public static void Shutdown()
        {
            RmlInstanceCache.Instance?.Clear();
            Native.Rml.Shutdown();
        }

        public static void SetSystemInterface(SystemInterface systemInterface)
        {
            Native.Rml.SetSystemInterface(systemInterface.NativePtr);
        }

        public static void SetRenderInterface(RenderInterface renderInterface)
        {
            Native.Rml.SetRenderInterface(renderInterface.NativePtr);
        }

        public static void SetFileInterface(FileInterface fileInterface)
        {
            Native.Rml.SetFileInterface(fileInterface.NativePtr);
        }

        /// <summary>
        /// Adds a new font face to the font engine. The face's family, style and weight will be determined from the face itself.
        /// </summary>
        /// <param name="fileName">The file to load the face from.</param>
        /// <param name="fallbackFace">True to use this font face for unknown characters in other font faces.</param>
        /// <param name="weight">The weight to load when the font face contains multiple weights, otherwise the weight to register the font as. By default it loads all found font weights.</param>
        /// <returns>True if the face was loaded successfully, false otherwise.</returns>
        public static bool LoadFontFace(string fileName, bool fallbackFace = false, FontWeight weight = FontWeight.Auto)
        {
            return Native.Rml.LoadFontFace(fileName, fallbackFace, weight);
        }

        public static EventId RegisterEventType(string type, bool interruptible, bool bubbles, DefaultActionPhase defaultActionPhase) {
            return Native.Rml.RegisterEventType(type, interruptible, bubbles, defaultActionPhase);
        }

        public static Context? CreateContext(string name, Vector2i dimensions, RenderInterface? renderInterface = null)
        {
            return Context.Create(name, dimensions, renderInterface);
        }

        public static bool RemoveContext(string name) {
            return Native.Rml.RemoveContext(name);
        }

        public static void RegisterPlugin(Plugin plugin) {
            Native.Rml.RegisterPlugin(plugin.NativePtr);
        }

        public static void Log(string message, LogType type = LogType.Always) {
            Native.Rml.Log(type, message);
        }

        public static RenderInterface GetRenderInterface() {
            return RenderInterface.Create(Native.Rml.GetRenderInterface());
        }
    }
}
